cmd/vm/switch_linux.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/sirupsen/logrus"
+	"github.com/songgao/packets/ethernet"
+	"github.com/songgao/water"
+	"github.com/vishvananda/netlink"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+)
+
+var (
+	debug    bool
+	tapIface string
+)
+
+const (
+	defaultTapDevice = "eth0"
+	defaultMacAddr   = "5a:94:ef:e4:0c:ee"
+	defaultMTU       = 4000
+)
+
+func main() {
+	flag.BoolVar(&debug, "debug", true, "enable debug flag")
+	flag.StringVar(&tapIface, "tap-interface", defaultTapDevice, "tap interface name, eg. eth0, eth1")
+	flag.Parse()
+
+	if debug {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+
+	// the FD is passed-in as an extra arg from exec.Command
+	// of the parent process. This is for the AF_VSOCK connection that
+	// is handed over from the default namespace to Rancher Desktop's
+	// network namespace (rd1), the logic behind this approach was the
+	// lack of support for network namespaces in the AF_VSOCK libs.
```

Comment:
Here is the exact wording from the open issue on vsock kernel module thread:
> _it could be nice to support network namespace in vsock_

Reply:
Well, the reason we need all thing fd-passing is because they ended up implementing network namespaces for AF_VSOCK (presumably in a commit later than that quoted bit); at least, that seems to be the observed behaviour (in that we don't get to use the same AF_VSOCK things inside the network namespace). That is, your comment here is exactly backwards — if AF_VSOCK didn't support namespaces, we wouldn't need to do this.